src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -1098,68 +1098,74 @@ class DataSetCache {
 public:
     DataSetCache() {}
 
-    /* The dataset is stored in a map, using the path of the file as the key and a vector of strings
-    with file's content as the value. In this method the empty vector is created and return its
-    pointer. getDataSetForPath method doesn't need the lock because loading and evaluating are
-    separate stages in Genny */
-    std::vector<std::string>* createMemoryForDataset(const std::string& path) {
-        std::lock_guard<std::mutex> lk(_dataset_mutex);
-        if (_all_datasets.count(path) > 0) {
-            /* When multiple threads try to load a dataset, all but the first thread will return
-               and do nothing. This way only the first thread will actually load the data.
-               Once the first vector is inserted and the first thread has finished running the
-               threads that were waiting on this mutex will find that there is already one element
-               for that _path and return a null pointer */
-            return nullptr;
-        }
+    using LoadFunction = std::function<void(std::vector<std::string>*)>;
+    /* The dataset is stored in a map, using the path of the file as the key and a Dataset object,
+    which contains the actual data as well as whether it is done being loaded,as the value.
+    In this method, the dataset at the given path is filled by the load function, and a reference
+    to the data is returned. getDataSetForPath method doesn't need the lock because loading and
+    evaluating are separate stages in Genny */
+    std::vector<std::string>& loadDataset(const std::string& path, LoadFunction loader) {
+        {
+            std::unique_lock<std::mutex> lk(_dataset_mutex);
```

Comment:
Move this outside & before this block. The loading thread also needs to hold the mutex before setting `done = true`, otherwise, the waiting threads can potentially miss a notify signal, and get stuck on `wait()`.

Reply:
good point. works for me!